x/gamm/keeper/migrate.go

```diff
@@ -52,19 +52,89 @@ func (k Keeper) MigrateFromBalancerToConcentrated(ctx sdk.Context, sender sdk.Ac
+		balancerToClPoolLink.ClPoolId = sdk.BigEndianToUint64(iter.Value())
+
+		balancerToClPoolLinks = append(balancerToClPoolLinks, balancerToClPoolLink)
+	}
+
+	migrationRecords := types.MigrationRecords{}
+	migrationRecords.BalancerToConcentratedPoolLinks = balancerToClPoolLinks
+	return migrationRecords, nil
+}
+
+// GetLinkedConcentratedPoolID returns the concentrated pool Id linked for the given balancer pool Id.
+// Returns error if link for the given pool id does not exist.
+func (k Keeper) GetLinkedConcentratedPoolID(ctx sdk.Context, balancerPoolId uint64) (uint64, error) {
+	store := ctx.KVStore(k.storeKey)
+	balancerToClPoolKey := types.GetKeyPrefixMigrationInfoBalancerPool(balancerPoolId)
+
+	concentratedPoolIdBigEndian := store.Get(balancerToClPoolKey)
+	if concentratedPoolIdBigEndian == nil {
+		return 0, types.ConcentratedPoolMigrationLinkNotFoundError{PoolIdLeaving: balancerPoolId}
+	}
+
+	return sdk.BigEndianToUint64(concentratedPoolIdBigEndian), nil
+}
+
+// GetLinkedConcentratedPoolID returns the Balancer pool Id linked for the given concentrated pool Id.
+// Returns error if link for the given pool id does not exist.
+func (k Keeper) GetLinkedBalancerPoolID(ctx sdk.Context, concentratedPoolId uint64) (uint64, error) {
+	store := ctx.KVStore(k.storeKey)
+	concentratedToBalancerPoolKey := types.GetKeyPrefixMigrationInfoPoolCLPool(concentratedPoolId)
+
+	balancerPoolIdBigEndian := store.Get(concentratedToBalancerPoolKey)
+	if balancerPoolIdBigEndian == nil {
+		return 0, types.ConcentratedPoolMigrationLinkNotFoundError{PoolIdLeaving: concentratedPoolId}
+	}
+
+	return sdk.BigEndianToUint64(balancerPoolIdBigEndian), nil
 }
 
 // SetMigrationInfo sets the balancer to gamm pool migration info to the store
+// Deletes all existing records, migrationInfo in state is completely overwitten by the given migrationInfo.
 func (k Keeper) SetMigrationInfo(ctx sdk.Context, migrationInfo types.MigrationRecords) {
 	store := ctx.KVStore(k.storeKey)
-	osmoutils.MustSet(store, types.KeyMigrationInfo, &migrationInfo)
+
+	// delete all existing keys
+	prefixStore := prefix.NewStore(store, types.KeyPrefixMigrationInfoBalancerPool)
+	iter := prefixStore.Iterator(nil, nil)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		prefixStore.Delete(iter.Key())
+	}
+
+	prefixStore = prefix.NewStore(store, types.KeyPrefixMigrationInfoCLPool)
+	iter = prefixStore.Iterator(nil, nil)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		prefixStore.Delete(iter.Key())
+	}
+
+	for _, balancerToCLPoolLink := range migrationInfo.BalancerToConcentratedPoolLinks {
+		balancerToClPoolKey := types.GetKeyPrefixMigrationInfoBalancerPool(balancerToCLPoolLink.BalancerPoolId)
+		store.Set(balancerToClPoolKey, sdk.Uint64ToBigEndian(balancerToCLPoolLink.ClPoolId))
+
+		clToBalancerPoolKey := types.GetKeyPrefixMigrationInfoPoolCLPool(balancerToCLPoolLink.ClPoolId)
+		store.Set(clToBalancerPoolKey, sdk.Uint64ToBigEndian(balancerToCLPoolLink.BalancerPoolId))
+	}
```

Comment:
Do set functions usually delete all existing keys? Feel like these two should potentially be separate functions (if you disagree, please feel free to resolve. Not blocking)

Reply:
Set function's purpose is to set the given object. Usually it overrides it automatically since set functions have a 1:1 relationship between the key and the object being stored.

This case is a special case, as we are having a 1:2 relationship. That is, within this set function, we are doing two different sets, causing the object being stored to not "automatically" over ride the object that was there before hand. That is why in this case, we manually iterate over the keys and delete all existing keys. 

I do think that what you are saying does make sense in terms of causing confusion, changing this to `OverrideAllMigrationInfo`